### PR TITLE
Simplify working with operands_and_partials

### PR DESCRIPTION
**Pull request by t4c1**
## Summary

Makes `broadcast_array` sum its argument when assigned to. When a scalar argument is broadcasted its derivative is always equal to sum of derivatives of equivalent vector argument. This simplifes working with `operands_and_partials` - no need for separate branches for scalar and vector arguments.

To use `sum` in `broadcast_array` I had to move it and `operands_and_partials` out of `meta` folder. I decided `functor` is the most appropriate. Due to this I had to change includes in many files.

`bernoulli_logit_lpmf` and `bernoulli_logit_glm_lpmf` have been simplified this way.

## Tests
Updated the test that checks how `broadcast_array` works when assigned a vector.

## Side Effects
I had to move some files out of mete folder - see summary.

## Release notes
Simplified using `operands_and_partials` by making `broadcast_array` sum the argument when assigned to.

## Checklist

- [x] Math issue #1470

- [x] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by t4c1** [comment]
This is ready for a review.

**Comment by t4c1** [comment]
Couls someone review this? Maybe @bob-carpenter, if you have time?

**Comment by bob-carpenter** [comment]
Sure.  I can review this one.

File: stan/math/prim/functor/broadcast_array.hpp [hunk 1]
```diff
@@ -23,14 +24,11 @@ class broadcast_array {
 
   /** \ingroup type_trait
    * We can assign any right hand side which allows for indexing to a
-   * broadcast_array. The idea is that the entry for the first index is what
-   * gets assigned. The most common use-case should be where the rhs is some
-   * container of length 1.
+   * broadcast_array. The idea is that the argument is summed.
```
Comment (bob-carpenter) [inline stan/math/prim/functor/broadcast_array.hpp:27]: [optional]
The original doc was very poorly written here.  Doc should not be in first person (plural or singular).  It should not say "The idea is that X"---it should just say X.  A compound like "use-case" should only be hyphenated when used as an adjective.

File: stan/math/prim/functor/broadcast_array.hpp [hunk 2]
```diff
@@ -23,14 +24,11 @@ class broadcast_array {
...
    */
   template <typename Y>
   void operator=(const Y& m) {
-    ref_type_t<Y> m_ref = m;
-    prim_ = m_ref[0];
+    prim_ = sum(m);
```
Comment (bob-carpenter) [inline stan/math/prim/functor/broadcast_array.hpp:31]: [question]
Does this cause any extra work if `m` is not a container?

Comment (t4c1) [reply stan/math/prim/functor/broadcast_array.hpp:31]: For non-container `sum` is a no-op. So assuming compiler is smart enough to inline the call, no.

**Comment by t4c1** [comment]
> No idea how the header tests passed before that.

`openrands_and_partials` was in meta folder, so it was included with `#include <stan/math/prim/meta.hpp>`

File: stan/math/prim/functor/broadcast_array.hpp (hunk 1 above)
Comment (t4c1) [reply stan/math/prim/functor/broadcast_array.hpp]: Fixed.

**Comment by t4c1** [comment]
@bob-carpenter This needs another review/approval.

**Pull request merged**